test/replication/qsync_advanced.test.lua

```diff
@@ -255,19 +263,20 @@ test_run:cmd('start server replica with wait=True, wait_load=True')
 _ = box.schema.space.create('sync', {engine=engine})
 _ = box.space.sync:create_index('pk')
 box.space.sync:insert{1} -- success
+test_run:wait_lsn('replica', 'default')
 test_run:cmd('switch replica')
 box.space.sync:select{} -- 1
 test_run:switch('default')
 -- Enable synchronous mode.
-disable_sync_mode()
+enable_sync_mode()
 -- Space is in sync mode now.
 box.cfg{replication_synchro_quorum=NUM_INSTANCES, replication_synchro_timeout=1000}
 box.space.sync:insert{2} -- success
-box.cfg{replication_synchro_quorum=BROKEN_QUORUM, replication_synchro_timeout=1000}
-box.space.sync:insert{3} -- success
-box.space.sync:select{} -- 1, 2, 3
+box.cfg{replication_synchro_quorum=BROKEN_QUORUM, replication_synchro_timeout=0.01}
+box.space.sync:insert{3} -- fail
+box.space.sync:select{} -- 1, 2
 test_run:cmd('switch replica')
-box.space.sync:select{} -- 1, 2, 3
+box.space.sync:select{} -- 1, 2
```

Comment:
I've got one failure here:
```
[007] Test failed! Result content mismatch:
[007] --- replication/qsync_advanced.result	Wed Apr 26 12:07:27 2023
[007] +++ /tmp/t/rejects/replication/qsync_advanced.reject	Wed Apr 26 12:09:00 2023
[007] @@ -773,6 +773,7 @@
[007]   | ---
[007]   | - - [1]
[007]   |   - [2]
[007] + |   - [3]
[007]   | ...
[007]  -- Testcase cleanup.
[007]  test_run:switch('default')
[007] 

```
It seems the `wait_lsn` is needed after the failed `insert{3}`. Because the ROLLBACK entry from master might not reach the replica in time.

Reply:
And one more. I don't understand how this could happen. The space is synchronous and quorum is 2 in that testcase:
```
[028] --- replication/qsync_advanced.result	Wed Apr 26 12:07:27 2023
[028] +++ /tmp/t/rejects/replication/qsync_advanced.reject	Wed Apr 26 12:17:25 2023
[028] @@ -93,7 +93,7 @@
[028]   | ...
[028]  box.space.sync:select{} -- 1
[028]   | ---
[028] - | - - [1]
[028] + | - []
[028]   | ...
[028]  -- Testcase cleanup.
[028]  test_run:switch('default')
[028] 
```

...
And another one:
```
[032] Test failed! Result content mismatch:
[032] --- replication/qsync_advanced.result	Wed Apr 26 12:07:27 2023
[032] +++ /tmp/t/rejects/replication/qsync_advanced.reject	Wed Apr 26 12:22:16 2023
[032] @@ -595,7 +595,7 @@
[032]   | ...
[032]  box.space.sync:select{} -- 1
[032]   | ---
[032] - | - - [1]
[032] + | - []
[032]   | ...
[032]  -- Testcase cleanup.
[032]  test_run:switch('default')
[032] 

```

I don't understand neither of these failures. In both cases the space is sync and quorum is 2.
Is it possible we caught an actual bug in limbo?

Reply:
one more update to make `default` ready after reconfig of replication.

Reply:
I now understand these magical failures with synchronous space and quorum 2:

<details><summary>1</summary>

```
[023] Test failed! Result content mismatch:
[004] replication/qsync_advanced.test.lua             vinyl           [ pass ]
[023] --- replication/qsync_advanced.result	Thu May  4 12:36:04 2023
[023] +++ /tmp/t/rejects/replication/qsync_advanced.reject	Thu May  4 12:47:03 2023
[023] @@ -93,7 +93,7 @@
[023]   | ...
[023]  box.space.sync:select{} -- 1
[023]   | ---
[023] - | - - [1]
[023] + | - []
[023]   | ...
[023]  -- Testcase cleanup.
[023]  test_run:switch('default')
[023] 
```
</details>
<details><summary>2</summary>

```
[021] replication/qsync_advanced.test.lua             vinyl           [ fail ]
[021] 
[021] Test failed! Result content mismatch:
[021] --- replication/qsync_advanced.result	Thu May  4 12:36:04 2023
[021] +++ /tmp/t/rejects/replication/qsync_advanced.reject	Thu May  4 12:51:04 2023
[021] @@ -520,7 +520,6 @@
[021]  box.space.sync:select{} -- 1, 2
[021]   | ---
[021]   | - - [1]
[021] - |   - [2]
[021]   | ...
[021]  -- Revert cluster configuration.
[021]  test_run:switch('default')
[021] 

```
</details>

It seems vinyl (and all the failures I caught were from vinyl) uses 'best-effort' transaction isolation, which's equal to 'read-confirmed' for read-only transactions.

So, you were right at the very beginning. We need to do `wait_lsn` after each `switch('replica')`.

Reply:
I propose another try with forcing txn_isolation on replica - both for vinyl and memtx.

Reply:
It seems `box.cfg{txn_isolation=...}` doesn't affect autocommit transactions. Like `box.space.sync:select{}`.

<details><summary>I propose the following fixup:</summary>

```diff
diff --git a/test/replication/qsync_advanced.result b/test/replication/qsync_advanced.result
index 947a4370b..067abac0c 100644
--- a/test/replication/qsync_advanced.result
+++ b/test/replication/qsync_advanced.result
@@ -66,14 +66,6 @@ test_run:cmd('start server replica with wait=True, wait_load=True')
 
 -- Successful write.
 -- Testcase setup.
--- Important to force vinyl to read committed, not confirmed
-test_run:switch('replica')
- | ---
- | - true
- | ...
-box.cfg{txn_isolation = 'read_committed'}
- | ---
- | ...
 test_run:switch('default')
  | ---
  | - true
@@ -99,7 +91,10 @@ test_run:switch('replica')
  | ---
  | - true
  | ...
-box.space.sync:select{} -- 1
+box.begin{txn_isolation='read-committed'} t = box.space.sync:select{} box.commit()
+ | ---
+ | ...
+t
  | ---
  | - - [1]
  | ...
@@ -191,7 +186,10 @@ test_run:switch('replica')
  | ---
  | - true
  | ...
-box.space.sync:select{} -- 1, 2, 3
+box.begin{txn_isolation='read-committed'} t = box.space.sync:select{} box.commit()
+ | ---
+ | ...
+t
  | ---
  | - - [1]
  |   - [2]
@@ -525,7 +523,10 @@ test_run:switch('default')
  | ---
  | - true
  | ...
-box.space.sync:select{} -- 1, 2
+box.begin{txn_isolation='read-committed'} t = box.space.sync:select{} box.commit()
+ | ---
+ | ...
+t
  | ---
  | - - [1]
  |   - [2]
@@ -604,7 +605,10 @@ test_run:switch('replica')
  | ---
  | - true
  | ...
-box.space.sync:select{} -- 1
+box.begin{txn_isolation='read-committed'} t = box.space.sync:select{} box.commit()
+ | ---
+ | ...
+t
  | ---
  | - - [1]
  | ...
diff --git a/test/replication/qsync_advanced.test.lua b/test/replication/qsync_advanced.test.lua
index db7983a45..179cf3dd0 100644
--- a/test/replication/qsync_advanced.test.lua
+++ b/test/replication/qsync_advanced.test.lua
@@ -31,9 +31,6 @@ test_run:cmd('start server replica with wait=True, wait_load=True')
 
 -- Successful write.
 -- Testcase setup.
--- Important to force vinyl to read committed, not confirmed
-test_run:switch('replica')
-box.cfg{txn_isolation = 'read_committed'}
 test_run:switch('default')
 box.cfg{replication_synchro_quorum=NUM_INSTANCES, replication_synchro_timeout=1000}
 _ = box.schema.space.create('sync', {is_sync=true, engine=engine})
@@ -42,7 +39,8 @@ box.ctl.promote()
 -- Testcase body.
 box.space.sync:insert{1} -- success
 test_run:switch('replica')
-box.space.sync:select{} -- 1
+box.begin{txn_isolation='read-committed'} t = box.space.sync:select{} box.commit()
+t
 -- Testcase cleanup.
 test_run:switch('default')
 box.space.sync:drop()
@@ -75,7 +73,8 @@ box.space.sync:insert{2}
 box.space.sync:insert{3}
 box.space.sync:select{} -- 1, 2, 3
 test_run:switch('replica')
-box.space.sync:select{} -- 1, 2, 3
+box.begin{txn_isolation='read-committed'} t = box.space.sync:select{} box.commit()
+t
 -- Testcase cleanup.
 test_run:switch('default')
 box.space.sync:drop()
@@ -190,7 +189,8 @@ box.cfg{replication_synchro_quorum = 2, replication_synchro_timeout = 1000}
 box.space.sync:insert{2}
 box.space.sync:select{} -- 1, 2
 test_run:switch('default')
-box.space.sync:select{} -- 1, 2
+box.begin{txn_isolation='read-committed'} t = box.space.sync:select{} box.commit()
+t
 -- Revert cluster configuration.
 test_run:switch('default')
 box.cfg{read_only=false}
@@ -216,7 +216,8 @@ box.space.sync:insert{2}
 box.error.injection.set('ERRINJ_WAL_IO', false)
 box.space.sync:select{} -- 1
 test_run:switch('replica')
-box.space.sync:select{} -- 1
+box.begin{txn_isolation='read-committed'} t = box.space.sync:select{} box.commit()
+t
 -- Testcase cleanup.
 test_run:switch('default')
 box.space.sync:drop()

```
</details>

With this diff applied, I see no failures locally.

Reply:
Applied, force pushed